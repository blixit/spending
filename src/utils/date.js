/**
 * Formats a date for the backend
 * @param {Date} date the date to convert
 */
export const toBackendDate = (date) => {
  return date
    ? date.toLocaleDateString() + ' ' + date.toLocaleTimeString()
    : date;
};