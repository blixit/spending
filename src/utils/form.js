/**
* Removes the listed fields from the source
* @param {*} form The source object
* @param {*} fields The fields list
*/
export const clean = (form, fields) => {
  const result = {};
  // eslint-disable-next-line no-unused-vars
  for (const field in form) {
    if (!form.hasOwnProperty(field) || fields.includes(field)) {
      continue;
    }
    result[field] = form[field];
  }
  return result;
};
