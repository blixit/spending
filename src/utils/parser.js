export const stringify = data => {
  try {
    return JSON.stringify(data);
  } catch (error) {
    return null;
  };
}

export const parse = data => JSON.parse(data);