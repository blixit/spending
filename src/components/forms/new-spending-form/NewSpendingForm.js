import React from 'react';
import styled from 'styled-components';
import trigger from 'vanillajs-browser-helpers/trigger';

import { mutate, ReadyMutations as Mutations } from 'core/http/query';
import { clean as cleanForm } from 'utils/form';
import { toBackendDate } from 'utils/date';

import Selector from 'components/inputs/selector/Selector';
import DatePickerComponent from 'components/inputs/date-picker/DatePicker';
import {
  NumberInput as NumberInputComponent,
  TextInput
} from 'components/inputs/input/Input';
import ThemedButton from 'components/atoms/buttons/ThemedButton';
import StyledError from 'components/molecules/error/Error';

const Form = styled.form`
  display: flex;
  flex-direction: column;
`;

const NumberInput = styled(NumberInputComponent)`
  width: 30%;
`;

const DatePicker = styled(DatePickerComponent)`
  margin-left: 10px;
`;

const Block = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  flex: 1 0 auto;
`;

class NewSpendingForm extends React.Component {
  categories = [
    'Restauration',
    'Taxis'
  ];
  state = {
    category: '',
    date: new Date(),
    price: null,
    name: 'My depense',
    hasError: false,
    errors: {}
  };

  onCategorySelected = (e) => {
    const category = e.target.value;
    this.setState({ category });
  };
  
  onDateChanged = (date) => this.setState({ date });
  
  onPriceChanged = (price) => this.setState({ price });
  
  onNameChanged = (name) => this.setState({ name });

  onSubmit = async (e) => {
    e.preventDefault();

    const { errors } = this.state;
    const { onFormResult } = this.props;
    const { add } = Mutations.spending;

    const body = cleanForm(this.state, ['hasError', 'errors']);

    try {
      const response = await mutate({
        ...add,
        data: {
          ...body,
          date: toBackendDate(body.date)
        }
      });
      onFormResult && onFormResult(response);
      this.setState({ hasError: false });
      trigger(document, 'refresh:balance', response);
    } catch (e) {
      const { message } = e.response.data || {};
      errors.global = message || add.queryError || e.message;
      this.setState({ hasError: true, errors });
    }
  };
  
  render () {
    const { date, price, name, hasError, errors } = this.state;
    const { onFormResult, ...rest } = this.props;
    return (
      <Form {...rest}>
        <Block>
          Category
          <Selector
            color='primary'
            values={this.categories}
            onChange={this.onCategorySelected}
          />
        </Block>
        <Block>
          Date
          <DatePicker popperPlacement="left-start" selected={date} onChange={this.onDateChanged} />
        </Block>
        <Block>
          Price
          <NumberInput value={price} onChange={this.onPriceChanged} />
        </Block>
        <Block>
          Name
          <TextInput value={name} onChange={this.onNameChanged} />
        </Block>
        <br />
        <ThemedButton color='primary' onClick={this.onSubmit}>Ajouter</ThemedButton>
        {hasError && <StyledError>{errors.global}</StyledError>}
      </Form>
    );
  }
}

export default NewSpendingForm;
