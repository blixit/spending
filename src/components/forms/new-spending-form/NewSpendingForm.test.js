import { withTheme } from 'ui/theming/test.utils';

import render from 'testing/render';

import NewSpendingForm from './NewSpendingForm';

describe('[components/forms/filter-form] NewSpendingForm', () => {
  it('should render properly', () => {
    const newSpendingForm = render(withTheme(NewSpendingForm, {}));
    expect(newSpendingForm.toJSON()).toMatchSnapshot();
  });
});
