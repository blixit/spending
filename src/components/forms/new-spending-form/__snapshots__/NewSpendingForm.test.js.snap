// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[components/forms/filter-form] NewSpendingForm should render properly 1`] = `
<form
  className="sc-jTzLTM iNxaQE"
>
  <div
    className="sc-ifAKCX bQstZT"
  >
    <label
      className="sc-bwzfXH cJnpOh sc-bxivhb jqRqDB"
    >
      <span>
        Date
      </span>
      <div
        className="react-datepicker-wrapper"
      >
        <div
          className="react-datepicker__input-container"
        >
          <input
            className="sc-bdVaJa fEZIYW"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="select a date"
            readOnly={false}
            type="text"
            value="18/10/2019"
          />
        </div>
      </div>
    </label>
  </div>
  <div
    className="sc-gzVnrw dnKKOo"
    placeholder="Ex: -45 for -45€"
    value={null}
  >
    <label
      className="sc-bwzfXH cJnpOh"
    >
      <span>
        Amount
      </span>
      <input
        className="sc-EHOje LArJT"
        onChange={[Function]}
        placeholder="Ex: -45 for -45€"
        type="number"
        value={null}
      />
    </label>
  </div>
  <div
    className="sc-gzVnrw dnKKOo"
    value="My depense"
  >
    <label
      className="sc-bwzfXH cJnpOh"
    >
      <span>
        Name
      </span>
      <input
        className="sc-EHOje LArJT"
        onChange={[Function]}
        type="text"
        value="My depense"
      />
    </label>
  </div>
  <div
    className="sc-iwsKbI LIYKF"
    color="primary"
  >
    <label
      className="sc-bwzfXH cJnpOh"
    >
      <span>
        Category
      </span>
      <select
        className="sc-htoDjs diwrEM"
        color="primary"
        onChange={[Function]}
      >
        <option>
          Restauration
        </option>
        <option>
          Taxis
        </option>
      </select>
    </label>
  </div>
  <br />
  <button
    className="sc-gZMcBi iltjcH sc-gqjmRU jMfTvD"
    onClick={[Function]}
  >
    Ajouter
  </button>
</form>
`;
